### PR TITLE
Add singlenode-ldap profile

### PR DESCRIPTION
**Pull request by sanjay990**
A new profile is created for testing ldap authentication.
Add basic infrastructure to support writing additional
product-test for LDAP authentication.

### CODE REVIEW

**Comment by sanjay990** [comment]
@arhimondr @kokosing - Can one of you review this.
@anusudarsan @amrutagokhale

File: presto-product-tests/bin/run.sh [hunk 1]
```diff
@@ -7,6 +7,12 @@ function absolutepath() {
     [[ $1 = /* ]] && echo "$1" || echo "$PWD/${1#./}"
 }
 
+# If profile is singlenode-ldap, import the root certificate of presto-coordinator
+if [ "$PROFILE" = "singlenode-ldap" ]
```
Comment (arhimondr) [inline presto-product-tests/bin/run.sh]: Could you please explain in details why it is needed?

File: presto-product-tests/conf/docker/common/standard.yml [hunk 2]
```diff
@@ -15,6 +15,18 @@ services:
     volumes:
       - ../../../../:/docker/volumes/presto
 
+  presto-master-ldap:
+    image: 'teradatalabs/centos6-java8-oracle-ldap'
+    hostname: presto-master
+    env_file:
+      - ../../../target/classes/presto.env
+    volumes:
+      - /etc/ldap/rootca
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/common/standard.yml]: What if my local operation system has no such catalog?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java [hunk 3]
```diff
@@ -127,6 +151,7 @@ public void shouldRunQuery()
             throws IOException, InterruptedException
     {
         launchPrestoCliWithServerArgument();
+        setLdapPassword();
```
Comment (anusudarsan) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: why do you have to call this method here instead of at say, line 210?

File: presto-product-tests/bin/run.sh (hunk 1 above)
Comment (sanjay990) [inline presto-product-tests/bin/run.sh]: In case of singlenode-ldap profile, we want the application runner ( container running cli ) to have the root certificate. Instead of creating a new container with that certificate, I am importing the certificate at run time for ldap-profile. I want to avoid creating a new docker image.  Also I can't import the certificate on the base image, as the base image is common for presto-master and application-runner (centos6-java8-oracle). We want only the root certificate on application-runner, not all certificates from presto-server.

File: presto-product-tests/conf/docker/common/standard.yml (hunk 2 above)
Comment (sanjay990) [inline presto-product-tests/conf/docker/common/standard.yml]: Sorry, didn't get that, which catalog. If you are talking about the volumes, that path exists in image.

**Comment by ArturGajowy** [comment]
How about testing this on travis?

**Comment by sanjay990** [comment]
@arhimondr : As of now, we can't test this on Travis. The user should have his own AD set up to run these tests (The LDAP specific tests which we would add in future) . We have plans to create docker-containers with openLDAP, post which can run on Travis. But since the priority is to have some product-tests, we moved with AD.

**Comment by ArturGajowy** [comment]
@sanjay990 That's fine as long as someone else ran the LDAP tests _and_ the openLDAP container is planned for the next sprint. Tests that aren't run rot and decay.

Also, notice that's another South Park avatar here ;)

**Comment by sanjay990** [comment]
@ArturGajowy : lol. Didn't see that. Sure, we will have tests running on Jenkins.

**Comment by sanjay990** [comment]
@arhimondr : Do you have any other comments.

File: presto-product-tests/conf/docker/common/standard.yml (hunk 2 above)
Comment (arhimondr) [inline presto-product-tests/conf/docker/common/standard.yml]: This line means that you want to mount `/etc/ldap/rootca` file or directory from the host machine to your docker container. For example my Ubuntu doesn't have such directory 

```
[andrii-thinkpad:~]  ⌘ ls /etc/ldap/rootca
ls: cannot access '/etc/ldap/rootca': No such file or directory
```

I'm not sure why it is needed, and whether it will work on any host maching. Anyway it looks suspicious.

File: presto-product-tests/conf/docker/common/standard.yml [hunk 4]
```diff
@@ -15,6 +15,18 @@ services:
     volumes:
       - ../../../../:/docker/volumes/presto
 
+  presto-master-ldap:
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/common/standard.yml]: Why do you need to add this section to the `standard.yml`? As i understand it is being used in `singlenode-ldap` profile only. Why not just move this section there?

File: presto-product-tests/bin/run.sh (hunk 1 above)
Comment (arhimondr) [inline presto-product-tests/bin/run.sh]: This script must be ready to be runned locally on the host OS. I don't want this script to modify keystores on my host OS. Why can't you just run the tempto in `centos6-java8-oracle-ldap` based container?

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  application-runner:
+    extends:
+      file: ../common/standard.yml
+      service: application-runner-common
+    depends_on:
+      - 'presto-master'
+    volumes_from:
+      - presto-master
+    environment:
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: use `teradatalabs/centos6-java8-oracle-ldap` image for this container and remove all the imperative stuffs. The same is done for kerberos.

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+discovery.uri=http://presto-master:8080
+
+# LDAP specific properties
+# https will have to enabled for ldap authentication
+http-server.https.port=8443
+http-server.https.enabled=true
+http-server.https.keystore.path=/etc/ldap/coordinator.jks
+http-server.https.keystore.key=testldap
+authentication.ldap.enabled=true
+authentication.ldap.url=ldaps://ad-test.presto.testldap.com:636
```
Comment (arhimondr) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: can't it be just `presto-master`? Why the ldap server is host specific?

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+authentication.ldap.ad-domain=presto.testldap.com
```
Comment (arhimondr) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties:27]: ditto?

P.S.: Take my appologies for the dumb question in advance. I have no idea how LDAP works internally.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java (hunk 3 above)
Comment (arhimondr) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: +1. Move it to the `launchPrestoCliWithServerArgument`

**Comment by anusudarsan** [comment]
also can you rebase this off sprint-35 branch? that should fix the travis build err.

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 6 above)
Comment (sanjay990) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: This has to be the url where the ldap server is running which in this case is AD running on openstack.
We are planning to use openLDAP for running these tests on Travis.

**Comment by sanjay990** [comment]
@arhimondr : Incorporated review comments. As suggested, I merged the coordinator and cli docker images.

**Comment by ArturGajowy** [comment]
@sanjay990 please use composition approach in the compose file for the new ldap profile, as in master and sprint-35. This way the compose file will be like 6 lines. See this PR for details: https://github.com/prestodb/presto/pull/5386/files.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java [hunk 8]
```diff
@@ -213,4 +257,13 @@ private void launchPrestoCli(List<String> arguments)
     {
         presto = new PrestoCliProcess(defaultJavaProcessLauncher().launch(Presto.class, arguments));
     }
+
+    private void setLdapPassword()
+    {
+        if (ldapAuthentication) {
```
Comment (anusudarsan) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: no need of this `if` condition here.

**Comment by sanjay990** [comment]
@ArturGajowy : Updated the docker-compose as per the new approach.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java (hunk 8 above)
Comment (sanjay990) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: Done

**Comment by ArturGajowy** [comment]
The docker part looks good to me

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,13 @@
+version: '2'
+services:
+
+  presto-master:
+    image: 'teradatalabs/centos6-java8-oracle-ldap'
+    command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode-ldap run
+    volumes:
+      - ../../../../presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml:/docker/volumes/tempto/tempto-configuration-local.yaml
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: Why do you need tempto configuration mounted to the presto-master image?

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    extra_hosts:
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: As i understand this service must be statically provisioned somewhere aside, so we can not run this profile on travis, right?

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+       - "ad-test.presto.testldap.com:10.25.171.180"
+
+  application-runner:
+    image: 'teradatalabs/centos6-java8-oracle-ldap'
```
Comment (arhimondr) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: Product tests are being run from this image? Shouldn't you mount the tempto configuration here?

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    host: presto-master
+    configured_hdfs_user: hive
+    cli_ldap_authentication: true
+    cli_ldap_truststore_path: /etc/ldap/cacerts.jks
+    cli_ldap_truststore_password: testldap
+    cli_ldap_user_name: financeuser1
+    cli_ldap_user_password: TCAMPass123
+    cli_ldap_server_address: https://${databases.presto.host}:8443
+
+  mysql:
```
Comment (arhimondr) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: We add those sections (`mysql`, `postgree`) in every single profile? They are always the same. Can't we just move it to the base tempto configuration which is located in product tests resources? If so please do such refactor in a separate commit, so we can push it to master futher without pushing other changes from this patch.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java [hunk 13]
```diff
@@ -167,15 +192,26 @@ public void shouldRunQueryFromFile()
...
+                    "--truststore-password", ldapTruststorePassword,
+                    "--user", ldapUserName,
+                    "--password");
+
             prestoClientOptions.add(arguments);
             launchPrestoCli(prestoClientOptions.build());
+            setLdapPassword();
         }
-        else {
+        else if (authentication) {
```
Comment (arhimondr) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/cli/PrestoCliTests.java]: As we have 2 types of authentication now it is better to rename this variable (and any similar to `kerberosAuthentication` and etc.)

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml (hunk 12 above)
Comment (sanjay990) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: I am of the same opinion. I will create a separate PR for that.

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml (hunk 10 above)
Comment (sanjay990) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: yes, that's correct. Once we have openLDAP in docker, we can remove this section and run tests on Travis.

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml (hunk 12 above)
Comment (arhimondr) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: That is gonna work to. Thanks.

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml (hunk 9 above)
Comment (sanjay990) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: I'll move it to application-runner

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml (hunk 12 above)
Comment (sanjay990) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: @arhimondr - https://github.com/Teradata/presto/pull/280

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml (hunk 11 above)
Comment (sanjay990) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: Done

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
+version: '2'
+services:
+
+  presto-master:
+    image: 'teradatalabs/centos6-java8-oracle-ldap'
+    command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode-ldap run
+    extra_hosts:
+       - "ad-test.presto.testldap.com:10.25.171.180"
```
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: Please change the hostname to something that won't ever be resolved to any domain squatter's domain (testldap.com is a nice domain if you're a bad guy). Also, please replace any references to Microsoft's concrete implementation (Active Directory) with LDAP. So replace it with e.g. `ldap`.

Comment (ArturGajowy) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: I'm also not sure if FB will merge a change that has a private IP hardcoded, so it might be this whole line will have to disappear and we won't be able to merge this before we have an AD docker container configured. Which shouldn't be that hard, there are plenty:  https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=ldap&starCount=0

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 7 above)
Comment (ArturGajowy) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties:27]: this is more of a username than an actual domain, right?

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 6 above)
Comment (ArturGajowy) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: ditto about the domain name

File: presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,13 @@
+version: '2'
+services:
+
+  presto-master:
+    image: 'teradatalabs/centos6-java8-oracle-ldap'
+    command: /docker/volumes/presto/presto-product-tests/conf/docker/files/presto-launcher-wrapper.sh singlenode-ldap run
+    extra_hosts:
+       - "${REPLACE_LDAP_SERVER_ADDRESS}:${REPLACE_LDAP_SERVER_IP}"
```
Comment (ArturGajowy) [inline presto-product-tests/conf/docker/singlenode-ldap/docker-compose.yml]: We misunderstood each other :) I meant the variables to be named without the 'replace' prefix + there should be an `environment` section, as below:

```
environment:
  LDAP_SERVER_ADDRESS: REPLACE_LDAP_SERVER_ADDRESS
  LDAP_SERVER_IP: REPLACE_LDAP_SERVER_IP
```

Or, if we want to avoid the repetition, the value of both vars can be 'REPLACE_ME'. So that user get's a 'no such host error: REPLACE_ME' if they don't specify the vars.

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 6 above)
Comment (ArturGajowy) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: This URL needs to change

**Comment by arhimondr** [comment]
lgtm

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 6 above)
Comment (sanjay990) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: As discussed changing this to ldap-server

**Comment by ArturGajowy** [comment]
LGTM :)
